moveit_core/online_signal_smoothing/include/moveit/online_signal_smoothing/acceleration_filter.h

```diff
@@ -0,0 +1,96 @@
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Paul Gesel
+   Description: Applies smoothing by limiting the  acceleration between consecutive commands
+ */
+
+#pragma once
+
+#include <cstddef>
+
+// Auto-generated
+#include <moveit_acceleration_parameters.hpp>
+#include <moveit/robot_model/robot_model.h>
+#include <moveit/robot_state/robot_state.h>
+#include <moveit/online_signal_smoothing/smoothing_base_class.h>
+
+namespace online_signal_smoothing
+{
+
+// Plugin
+class AccelerationLimitedPlugin : public SmoothingBaseClass
+{
+public:
+  /**
+   * Initialize the acceleration based smoothing plugin
+   * @param node ROS node, used for parameter retrieval
+   * @param robot_model typically used to retrieve vel/accel/jerk limits
+   * @param num_joints number of actuated joints in the JointGroup Servo controls
+   * @return True if initialization was successful
+   */
+  bool initialize(rclcpp::Node::SharedPtr node, moveit::core::RobotModelConstPtr robot_model,
+                  size_t num_joints) override;
+
+  /**
+   * Smooth the command signals for all DOF. This function estimates the velocity and acceleration using the two
+   * previous positions passed in. It is designed to be called in a control loop at a constant interval.
+   * @param positions array of joint position commands
+   * @param velocities (unused)
+   * @param accelerations (unused)
+   * @return True if smoothing was successful
+   */
+  bool doSmoothing(Eigen::VectorXd& positions, Eigen::VectorXd& velocities, Eigen::VectorXd& accelerations) override;
+
+  /**
+   * Reset to a given joint state. This method must be called before doSmoothing.
+   * @param positions reset the filters to these joint positions
+   * @param velocities (unused)
+   * @param accelerations (unused)
+   * @return True if reset was successful
+   */
+  bool reset(const Eigen::VectorXd& positions, const Eigen::VectorXd& velocities,
+             const Eigen::VectorXd& accelerations) override;
+
+private:
+  rclcpp::Node::SharedPtr node_;
+  long num_joints_;
+  std::array<std::pair<Eigen::VectorXd, std::optional<rclcpp::Time>>, 2> last_positions_;
```

Comment:
I think this array requires some more explanation, as I got a little confused about whether its size had to do with the joints vs. something else.

As I now understand, we need a buffer of 2 previous positions to actually limit the acceleration, right? If so, explain that and also what the 0th and 1st indices correspond to.

Also, please include in the comments that the pair is the actual position and its corresponding time, if one is available. I know this part is clear(ish) in the actual type, but just to be thorough.